### PR TITLE
Update on theory.md and examples.

### PR DESCRIPTION
**Pull request by HesamShaelaie**
This is not complete and still needs more work.

### CODE REVIEW

File: docs/src/references.bib [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+@article{dempe2015bilevel,
+  title={Bilevel programming problems},
+  author={Dempe, Stephan and Kalashnikov, Vyacheslav and P{\'e}rez-Vald{\'e}s, Gerardo A and Kalashnykova, Nataliya},
+  journal={Energy Systems. Springer, Berlin},
+  year={2015},
+  publisher={Springer}
+}
+
+<a id="6">[6]</a> 
+Dempe, Stephan, and Alain Zemkoho. Bilevel Optimization. Springer, 2020.
```
Comment (odow) [inline docs/src/references.bib:57]: Remove these

File: docs/src/references.bib [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+@book{dempe2020bilevel,
+  title={Bilevel Optimization},
+  author={Dempe, Stephan and Zemkoho, Alain},
+  year={2020},
+  publisher={Springer}
+}
+
+<a id="7">[7]</a> 
+Fortuny-Amat, José, and Bruce McCarl. "A representation and economic interpretation of a two-level programming problem." Journal of the operational Research Society 32.9 (1981): 783-792.
```
Comment (odow) [inline docs/src/references.bib:67]: And here. And so on

File: docs/src/examples/FOBP_example4.jl [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# Foundations of Bilevel Programming: Example 2.1
```
Comment (odow) [inline docs/src/examples/FOBP_example4.jl:1]: We should delete these files and add them a different Pull Request.

File: docs/src/examples/FOBP_example3.jl [hunk 4]
```diff
@@ -0,0 +1,77 @@
+# Foundations of Bilevel Programming: Example 3
```
Comment (odow) [inline docs/src/examples/FOBP_example3.jl:1]: Delete

File: docs/src/examples/FOBP_example2.2.jl [hunk 5]
```diff
@@ -1,29 +1,26 @@
-# # Foundations of Bilevel Programming: Example 2.1
```
Comment (odow) [inline docs/src/examples/FOBP_example2.2.jl:1]: What happened here? We don't want to modify this file in this PR. Do it in a different one.

File: docs/src/examples/FOBP_example2.1.jl [hunk 6]
```diff
@@ -0,0 +1,107 @@
+# Foundations of Bilevel Programming: Example 2.1
```
Comment (odow) [inline docs/src/examples/FOBP_example2.1.jl:1]: Add this in a different PR.

File: .gitignore [hunk 7]
```diff
@@ -5,5 +5,7 @@ deps/deps.jl
 Manifest.toml
 docs/build
 docs/src/examples/*.md
-docs/src/.DS_Store
 .DS_Store
+*/.DS_Store
+*/*/.DS_Store
+docs/src/.DS_Store
```
Comment (odow) [inline .gitignore:11]: We need to delete the `docs/.DS_Store` file and add it to `.gitignore`.

File: docs/src/theory.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+2. The set ``\{(x,y): x \in \Omega_x, y\in \Psi(x) \} `` denotes the optimal follower solutions lifted to the x-y-space, and is given by the green and red facets. This set is nonconvex.
+3. The single leader constraint indicated by the dashed line renders certain optimal responses of the follower infeasible. Thus, the bilevel feasible region ``\mathcal{F}`` corresponds to the green facets. Consequently, the feasible set of example (1) is not only nonconvex but also disconnected.
+4. The optimal solution of example (1) is (3/7, 6/7) with objective function value 39/7. In contrast, ignoring the follower objective, i.e., solving the high-point relaxation (HPR), yields the optimal solution (3, 0) with objective function value 3. Note that the latter point is not bilevel feasible.
+
+
+Example (1) indicates that bilevel problems are intrinsically nonconvex due to their hierarchical structure. It has been proven that LP-LP problems are strongly NP-hard; even checking local optimality of LP-LP problems is NP-hard. 
+
+
+##  Linear and Convex Follower Problems
```
Comment (odow) [inline docs/src/theory.md:181]: I think we can remove this section

File: docs/src/theory.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
...ith convex follower problems today. Alternative approaches, e.g., penalty methods or descent approaches, have been proposed by Anandalingam and White [anandalingam1990solution](@cite) and Savard and Gauvin [savard1994steepest](@cite), respectively. In the 1990s, the largest instances of linear bilevel problems that have been solved consisted of 250 leader variables, 150 follower variables, and 150 follower constraints; see Hansen et al. [hansen1992new](@cite). Although cutting planes have been derived in the following years, see Audet, Haddad, et al. [audet2007disjunctive](@cite) and Audet, Savard, et al. [audet2007new](@cite), computational linear and convex bilevel optimization did not attract much attention in the 2000s and not many computational results have been reported. Similarly, Moore and Bard [candler1977multi](@cite) developed a branch-and-bound approach for bilevel problems with mixed-integer follower problems and also reported some first numerical results already in 1990. However, only very little computational progress has been reported until DeNegre and Ralphs [denegre2009branch](@cite) introduced a branch-and-cut approach for purely integer bilevel problems in 2009.
```
Comment (odow) [inline docs/src/theory.md:35]: This section can be removed.

**Review by odow — COMMENTED** [review]
Delete the example files, `docs/.DS_Store`, and delete some of the sections in `theory.md`. 

Then we can merge and do any improvements in a future PR.

**Comment by odow** [comment]
@HesamShaelaie you'll need to fix the conflict like I showed you before.

Then on your local machine, run 
```
git checkout UpdateDocumentation
git pull
```
to get the new change.

If you want to resolve the conflict locally it's a might more opaque. Try
```
git checkout master
git pull
git checkout UpdateDocumentation
git rebase master
git status  # You should see some things in red. 
# Go edit the offending files. Once you're done
git add .
git rebase --continue
git push -f  # The `-f` is okay here
```

**Review by joaquimg — COMMENTED** [review]
I added some notes.

We should re-think this section. We should not have any copy of Thomas' Thesis. It would violate copyright.

Using the same notation and highlighting that is where you got the same notation is great. And instead of copying, we can precisely point to the section in the thesis.

One way forward is:
Decide which are the main concepts we want to explain, list them explain them very briefly, possibly by example (not necessary though).
Point to references for further reading.

We should not try to teach all about bilevel here.
I believe we should attempt to highlight: formulations, concepts, and definitions that are used throughout the package (mostly because there are some variations).

**Comment by odow** [comment]
Going to close this for now. We can come back to it later.

**Pull request closed without merging**